module/jobqueue/consumer.go

```diff
@@ -0,0 +1,310 @@
+func (c *Consumer) run() (int64, error) {
+	if !c.running {
+		return 0, nil
+	}
+
+	processedFrom := c.processedIndex
+	processables, processedTo, err := c.processableJobs()
+	if err != nil {
+		return 0, fmt.Errorf("could not query processable jobs: %w", err)
+	}
+
+	c.log.Debug().
+		Int64("processed_from", processedFrom).
+		Int64("processed_to", processedTo).
+		Int("processables", len(processables)).
+		Msg("running")
+
+	for _, jobAtIndex := range processables {
+		jobID := jobAtIndex.job.ID()
+
+		c.processingsIndex[jobID] = jobAtIndex.index
+		c.processings[jobAtIndex.index] = &jobStatus{
+			jobID: jobID,
+			done:  false,
+		}
+
+		go c.worker.Run(jobAtIndex.job)
+	}
+
+	err = c.progress.SetProcessedIndex(processedTo)
+	if err != nil {
+		return 0, fmt.Errorf("could not set processed index %v, %w", processedTo, err)
+	}
+	c.processedIndex = processedTo
+
+	return int64(len(processables)), nil
+}
+
+func (c *Consumer) processableJobs() ([]*jobAtIndex, int64, error) {
+	return processableJobs(
+		c.jobs,
+		c.processings,
+		c.maxProcessing,
+		c.maxFinished,
+		c.processedIndex,
+	)
+}
+
+// processableJobs check the worker's capacity and if sufficient, read
+// jobs from the storage, return the processable jobs, and the processed
+// index
+func processableJobs(jobs storage.Jobs, processings map[int64]*jobStatus, maxProcessing int64, maxFinished int64, processedIndex int64) ([]*jobAtIndex, int64, error) {
+	processables := make([]*jobAtIndex, 0)
+
+	// count how many jobs are still processing,
+	// in order to decide whether to process a new job
+	processing := int64(0)
+	finished := int64(0)
+
+	// if still have processing capacity, find the next processable job
+	for i := processedIndex + 1; processing < maxProcessing && finished < maxFinished; i++ {
+		status, ok := processings[i]
+
+		// if no worker is processing the next job, try to read it and process
+		if !ok {
+			// take one job
+			job, err := jobs.AtIndex(i)
+
+			// if there is no more job at this index, we could stop
+			if errors.Is(err, storage.ErrNotFound) {
+				break
+			}
+
+			// exception
+			if err != nil {
+				return nil, 0, fmt.Errorf("could not read job at index %v, %w", i, err)
+			}
+
+			processing++
+
+			processables = append(processables, &jobAtIndex{
+				job:   job,
+				index: i,
+			})
+			continue
+		}
+
+		// only increment the processing variable when
+		// the job is not done, meaning still processing
+		if !status.done {
+			processing++
+			continue
+		}
+
+		if i == processedIndex+1 {
+			processedIndex++
+		} else {
+			finished++
+		}
+	}
+
+	return processables, processedIndex, nil
+}
+
+// doneJob updates the internal state to mark the job has been processed
+func (c *Consumer) doneJob(jobID JobID) bool {
+	// lock
+	index, ok := c.processingsIndex[jobID]
+	if !ok {
+		// job must has been processed
+		return false
+	}
+
+	status, ok := c.processings[index]
+	if !ok {
+		// must be a bug, if went here
+		return false
```

Comment:
This better to be distinguished than the previous `if` case. I would return an error since it would indicate a leakage that violates the atomicity. With just a single returned value it is hard to debug.

Reply:
what is the behaviour on a fatal level log, does it exit the program?

Reply:
Added log to crash the node if this ever happen
https://github.com/onflow/flow-go/blob/da5a24b04d481d4dd59d4269bda6be54e4078a54/module/jobqueue/consumer.go#L288